lib/rspec/mocks/proxy.rb

```diff
@@ -98,10 +98,19 @@ def replay_received_message_on(expectation, &block)
 
       # @private
       def check_for_unexpected_arguments(expectation)
-        @messages_received.each do |(method_name, args, _)|
-          next unless expectation.matches_name_but_not_args(method_name, *args)
+        # Return unless message never received any of the specified args
+        return unless all_arguments_unexpected?(expectation)
+        unexpected_args = @messages_received.map { |_method_name, args, _| args }.uniq.flatten
 
-          raise_unexpected_message_args_error(expectation, *args)
+        raise_unexpected_message_args_error(expectation, unexpected_args)
+      end
+
+      # @private
+      def all_arguments_unexpected?(expectation)
```

Comment:
I'd think I'd prefer we do this inline, in the method above e.g:

```Ruby
def check_for_unexpected_arguments(expectation)
  return if @messages_received.empty?
  return if @messages_received.none? do |(method_name, args, _)|
              expectation.matches_name_but_not_args(method_name, *args)
            end
  # ...
end
```

I switched the `all?` for `none?` rather than an unless, I think thats the same semantics...